### PR TITLE
extend and clean up the SimulatorReport

### PR DESCRIPTION
**Pull request by andlaus**
this is the compagnion PR to OPM/opm-core#1112. I've made sure that everything still compiles and that no additional unit tests fail. I've also checked that `flow_legacy` and `flow_ebos` work as expected for the Norne deck:

```
# tail -n 20 flow_ebos.log flow_legacy.log 
==> flow_ebos.log <==
[...]
Total time taken (seconds):   761.545
Solver time (seconds):        745.516
 Assembly time (seconds):     355.981
 Linear solve time (seconds): 319.673
 Update time time (seconds):  2.69076
 Output write time (seconds): 16.9406
Overall Linearizations:       1964
Overall Newton Iterations:    1621
Overall Linear Iterations:    24638


==> flow_legacy.log <==
[...]
Total time taken (seconds):   1259.69
Solver time (seconds):        1253.5
 Assembly time (seconds):     701.505
 Linear solve time (seconds): 352.299
 Update time time (seconds):  35.5654
 Output write time (seconds): 1.1171
Overall Linearizations:       1973
Overall Newton Iterations:    1610
Overall Linear Iterations:    25265
```

these results indicate that the linearization in `flow_ebos` is about twice as fast as that of `flow_legacy` and that the linear solver is approximately the same once the results have been adjusted for the different number of linar solves.

### CODE REVIEW

**Comment by bska** [comment]
>
```
 Update time time (seconds):
```

Is this really supposed to say `time time`?

**Comment by blattms** [comment]
What is "update time"?

**Comment by blattms** [comment]
Out of curiosity: Any idea why the output write is do significantly different (factor 15)?

**Comment by andlaus** [comment]
> What is "update time"?

the time needed for doing `x_new = x_old - dx` in the Newton update. since this involves the ReservoirState and thus a lot of non-trivial code, I thought that this number should be included...

**Comment by andlaus** [comment]
> Out of curiosity: Any idea why the output write is do significantly different (factor 15)?

possibly a bug.

**Comment by andlaus** [comment]
> possibly a bug.

hm, if it is a bug it  is a non-obvious one: the Simulator::run() from `SimulatorBase_impl.hpp` and `SimulatorFullyImplicitBlackoil.hpp` seem to do exactly the same w.r.t. this.

**Comment by blattms** [comment]
Are you sure? Then I wonder why there are files called SimulatorFullyImplicitBlackoilOuputEbos.[hc]pp.

**Comment by andlaus** [comment]
> Are you sure?

pretty much. I just remembered: in `flow_legacy` creating the arrays for `opm-output` comes down to copying the `value` member of the respective `ADB` while it is not as trivial in `flow_ebos`.

Anyway, even for `flow_ebos` output takes still less than 2.5 % of the solver time, so it is not really relevant IMO. What's more interesting is the 50.23 seconds "missing" time when you subtract all constituents from the solver time. I suppose this is caused by some unfortunate design decisions like creating the model for every single time step, but I actually have not looked into this more deeply.

**Comment by andlaus** [comment]
jenkins run norne_ebos OPM/opm-core=1112 please

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 1]
```diff
@@ -264,16 +270,26 @@ namespace Opm {
...
+            }
+            catch (...) {
+                report.assemble_time += perfTimer.stop();
+                throw; // continue throwing the stick
+            }
 
-            IterationReport iter_report = assemble(timer, iteration, reservoir_state, well_state);
             std::vector<double> residual_norms;
-            const bool converged = getConvergence(timer, iteration,residual_norms);
+            report.converged = getConvergence(timer, iteration,residual_norms);
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: Not measured.

**Comment by andlaus** [comment]
jenkins run norne OPM/opm-core=1112 please

**Comment by andlaus** [comment]
jenkins run norne_ebos opm-core=1112 please

**Comment by andlaus** [comment]
jenkins run norne opm-core=1112 please

File: opm/autodiff/BlackoilModelBase_impl.hpp [hunk 2]
```diff
@@ -250,17 +256,31 @@ typedef Eigen::Array<double,
...
         }
-        IterationReport iter_report = asImpl().assemble(reservoir_state, well_state, iteration == 0);
+        asImpl().assemble(reservoir_state, well_state, iteration == 0);
         residual_norms_history_.push_back(asImpl().computeResidualNorms());
-        const bool converged = asImpl().getConvergence(timer, iteration);
-        const bool must_solve = (iteration < nonlinear_solver.minIter()) || (!converged);
+
+        report.assemble_time += perfTimer.stop();
+        report.total_linearizations = 1;
+        report.converged = asImpl().getConvergence(timer, iteration);
```
Comment (andlaus) [inline opm/autodiff/BlackoilModelBase_impl.hpp]: seems like the time for this is about 6 seconds for Norne on my machine:

```
Total time taken (seconds):   774.747
Solver time (seconds):        758.785
 Assembly time (seconds):     366.692
 Linear solve time (seconds): 320.887
 Update time (seconds):       8.64163
 Output write time (seconds): 16.9406
Overall Linearizations:       1965
Overall Newton Iterations:    1622
Overall Linear Iterations:    24698
```

**Comment by bska** [comment]
jenkins run spe norne polymer opm-core=1112 please

**Comment by totto82** [comment]
Is total_time - solver_time =  parsing_time + setup_time?

**Comment by andlaus** [comment]
I think yes, but I'm not 100 sure: It is the "overhead time", before the actual simulation starts (whatever that is, there might be something more in it).

**Comment by andlaus** [comment]
seems like jenkins has some unrelated issues:

```
Program threw an exception: boost::filesystem::canonical: No such file or directory: "/var/lib/jenkins/workspace/opm-simulators-PR-builder/deps/opm-data/SPE1CASE2.DATA"
```

**Comment by akva2** [comment]
sorry, my bad. i had run some tests locally to debug some test issues, and did so using the wrong user. fixed now.

**Comment by atgeirr** [comment]
jenkins run spe norne polymer opm-core=1112 please

**Comment by andlaus** [comment]
happy jenkins; good jenkins

File: opm/autodiff/BlackoilModelBase_impl.hpp [hunk 3]
```diff
@@ -757,17 +780,16 @@ typedef Eigen::Array<double,
...
+            return;
         }
 
         std::vector<ADB> mob_perfcells;
         std::vector<ADB> b_perfcells;
         asImpl().wellModel().extractWellPerfProperties(state, sd_.rq, mob_perfcells, b_perfcells);
         if (param_.solve_welleq_initially_ && initial_assembly) {
             // solve the well equations as a pre-processing step
-            iter_report = asImpl().solveWellEq(mob_perfcells, b_perfcells, reservoir_state, state, well_state);
+            asImpl().solveWellEq(mob_perfcells, b_perfcells, reservoir_state, state, well_state);
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelBase_impl.hpp]: This is a significant change, if the pre-well-solve goes bad we should not lose that info.

File: opm/simulators/WellSwitchingLogger.cpp [hunk 4]
```diff
@@ -137,7 +137,7 @@ void WellSwitchingLogger::logSwitch(const char* name, std::array<char,2> fromto,
                                     int rank)
 {
             std::ostringstream ss;
-            ss << "Switching control mode for well " << name
+            ss << "    Switching control mode for well " << name
```
Comment (atgeirr) [inline opm/simulators/WellSwitchingLogger.cpp]: This seems unrelated to me?

File: opm/autodiff/BlackoilModelBase.hpp [hunk 5]
```diff
@@ -175,7 +176,7 @@ namespace Opm {
         /// \param[in, out] reservoir_state   reservoir state variables
         /// \param[in, out] well_state        well state variables
         template <class NonlinearSolverType>
-        IterationReport nonlinearIteration(const int iteration,
+        SimulatorReport nonlinearIteration(const int iteration,
```
Comment (andlaus) [inline opm/autodiff/BlackoilModelBase.hpp]: In the medium term I intend to do so. It would also be good for the fully implicit solvers to account for failed time steps; come think about it, that's probably the bulk of the "missing" solver time. the only obstacle for this which I can come up with of out of my head is that access to the report object must be made exception-proof for this.

File: opm/autodiff/BlackoilModelBase_impl.hpp [hunk 6]
```diff
@@ -50,8 +50,10 @@
 #include <opm/core/utility/parameters/ParameterGroup.hpp>
 #include <opm/parser/eclipse/EclipseState/EclipseState.hpp>
 #include <opm/parser/eclipse/EclipseState/Tables/TableManager.hpp>
-
 #include <opm/common/data/SimulationDataContainer.hpp>
+
+#include <dune/common/timer.hh>
```
Comment (andlaus) [inline opm/autodiff/BlackoilModelBase_impl.hpp]: ok. the reason why I opted for `Dune::Timer` instead of `Opm::StopWatch` was that `opm-core` is scheduled to be dissolved?

File: opm/simulators/WellSwitchingLogger.cpp (hunk 4 above)
Comment (andlaus) [inline opm/simulators/WellSwitchingLogger.cpp]: yeah, but that makes the terminal output quite a bit less cluttered.

File: opm/autodiff/BlackoilModelBase_impl.hpp (hunk 3 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelBase_impl.hpp]: true. changed.

**Comment by andlaus** [comment]
jenkins run spe norne norne_ebos polymer opm-core=1112 please

**Comment by andlaus** [comment]
hm: did jenkins just run norne twice with flow_ebos?

@akva2 bug? did I tell it to do something which I wasn't supposed to?

**Comment by akva2** [comment]
Due to inherited design (job runs directly in opm-data norne dir) you cannot do both runs in one go. I havent found it important enough to prioritize fixing this.

File: opm/autodiff/BlackoilModelBase.hpp (hunk 5 above)
Comment (atgeirr) [inline opm/autodiff/BlackoilModelBase.hpp]: ... but adding the "failed" member (now in IterationReport) should not require any particular exception proofing, right?

File: opm/simulators/WellSwitchingLogger.cpp (hunk 4 above)
Comment (atgeirr) [inline opm/simulators/WellSwitchingLogger.cpp]: Good, just wanted to check there was no deeper point I was missing.

**Comment by atgeirr** [comment]
> Due to inherited design (job runs directly in opm-data norne dir) you cannot do both runs in one go. I havent found it important enough to prioritize fixing this.

If the problem is output cluttering using "output_dir=something" should help, unless it is more complicated than I realize?

**Comment by akva2** [comment]
you also have to change the scripts that generate plots etc. but yes, that is basically what is required as such, then you deal with the consequences..

File: opm/autodiff/BlackoilModelBase.hpp (hunk 5 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelBase.hpp]: no, but that's also not the problem: the problem is accessing the report object after an exception is thrown because nothing can returned to the caller.

Comment (atgeirr) [inline opm/autodiff/BlackoilModelBase.hpp]: Perhaps we are not quite aligned as to how we want to use this member, but for sure: we must stop throwing if we are to propagate such info back to the caller. I am not proposing to do that with this PR though!

Comment (andlaus) [inline opm/autodiff/BlackoilModelBase.hpp]: IMO this method is unfortunate because it basically duplicates the functionality of the report (and always has been) which will lead to inconsistencies or at least confusion.

maybe the report should be stored as an attributed and accessed via a `lastReport()` member?

Comment (andlaus) [inline opm/autodiff/BlackoilModelBase.hpp]: simple: on one side, there is the report which is returned by the `step()` method, on the other side, there are `numIterations()` and friends which are completely orthogonal and must be used in case of exceptions...
